File: examples/ios/CMeSdkObjc/CMeSdkObjc/HomeViewController.m
//
//  HomeViewController.m
//  CMeSdkObjc
//
//  Created by Norman Jarvis on 5/3/19.
//  Copyright © 2019 Evernym Inc. All rights reserved.
//
#import "CustomTableViewCell.h"
#import "HomeViewController.h"
#import "MobileSDK.h"
#import "CMConnection.h"
#import "CMCredential.h"
#import "CMUtilities.h"
#import "LocalStorage.h"
#import "QRCodeReaderViewController.h"
#import "CMMessage.h"
#import "CMProofRequest.h"

@interface HomeViewController ()

@property (weak, nonatomic) IBOutlet UILabel *infoLbl;
@property (nonatomic, readwrite, strong) IBOutlet UITableView *tableView;

@property NSDictionary* requests;
@property BOOL isInitialized;

@end

@implementation HomeViewController
@synthesize addConnLabel, addConnConfigTextView, requests;

UIGestureRecognizer *tapper;

- (void)viewDidLoad {
    [super viewDidLoad];
    self.tableView.dataSource = self;
    self.tableView.delegate = self;
    addConnConfigTextView.delegate = self;
    addConnConfigTextView.layer.cornerRadius = 5;
}

- (void)viewWillAppear:(BOOL)animated {
    [super viewWillAppear: animated];
    requests = [LocalStorage getObjectForKey: @"requests" shouldCreate: false];
    [self.tableView reloadData];
    [[NSNotificationCenter defaultCenter] addObserver: self selector: @selector(vcxInitialized) name:@"vcxInitialized" object: nil];
    _isInitialized = [[MobileSDK shared] sdkInited];
}

- (void) viewWillDisappear:(BOOL)animated {
    [[NSNotificationCenter defaultCenter] removeObserver: self];
}

- (void)handleSingleTap:(UITapGestureRecognizer *) sender {
    [self.view endEditing: YES];
}

- (void)vcxInitialized {
    self.infoLbl.text = @"VCX initialized!";
    _isInitialized = true;
    double delayInSeconds = 10.0;
    dispatch_time_t popTime = dispatch_time(DISPATCH_TIME_NOW, (int64_t)(delayInSeconds * NSEC_PER_SEC));
    dispatch_after(popTime, dispatch_get_main_queue(), ^(void){
        self.infoLbl.text = @"";
    });
}

- (BOOL)textViewShouldBeginEditing:(UITextView *)textView {
    if([textView.text isEqual: @"enter code here"]) {
        addConnConfigTextView.text = @"";
    }
    return true;
}

- (IBAction)scanQR: (UIButton*) sender {
    if(!_isInitialized) {
        NSLog(@"Please wait for VCX to initialize!");
        return;
    }

    QRCodeReader *reader = [QRCodeReader readerWithMetadataObjectTypes:@[AVMetadataObjectTypeQRCode]];

    QRCodeReaderViewController *vc = [QRCodeReaderViewController readerWithCancelButtonTitle:@"Cancel" codeReader: reader startScanningAtLoad:YES showSwitchCameraButton:YES showTorchButton:YES];
    vc.modalPresentationStyle = UIModalPresentationFormSheet;

    [self presentViewController: vc animated: YES completion:^{
        NSLog(@"QR code scanner presented");
    }];

    [reader setCompletionWithBlock:^(NSString *scanResult) {
        NSLog(@"%@", scanResult);
        self.addConnConfigTextView.text = scanResult;
        [self addNewConn: sender];
        [self dismissViewControllerAnimated: vc completion:^{
            NSLog(@"QR code scanner dissmised");
        }];
    }];
}

- (void)newConnection:(NSString *) data
  withCompletionBlock:(ResponseWithBoolean) completionBlock {
    NSLog(@"newConnection %@", data);
    [CMConnection handleConnection:data
                    connectionType:QR
                       phoneNumber:@""
             withCompletionHandler:^(NSDictionary *responseObject, NSError *error) {
        if (error && error.code > 0) {
            return completionBlock(NO, error);
        }
        return completionBlock(YES, error);
    }];
}

- (void)answer:(NSString *) data
withCompletionBlock:(ResponseWithBoolean) completionBlock {
    NSDictionary *message = [CMUtilities jsonToDictionary:data];
    NSString *payload = [message objectForKey:@"payload"];
    NSDictionary *payloadDict = [CMUtilities jsonToDictionary:payload];
    NSArray *responses = [payloadDict objectForKey:@"valid_responses"];

    NSString *pwDidMes = [message objectForKey:@"pwDid"];
    NSString *questionsConnection = [CMConnection getConnectionByPwDid:pwDidMes];

    UIAlertController * alert = [UIAlertController
                                     alertControllerWithTitle:[payloadDict objectForKey:@"question_text"]
                                     message:[payloadDict objectForKey:@"question_detail"]
                                     preferredStyle:UIAlertControllerStyleAlert];
    
    for(NSInteger i = 0; i < responses.count; i++) {
        NSDictionary *response = responses[i];
        UIAlertAction* button = [UIAlertAction
                                actionWithTitle:[response objectForKey:@"text"]
                                style:UIAlertActionStyleDefault
                                handler:^(UIAlertAction * action) {
                                    [CMMessage answerQuestion:questionsConnection
                                                 message:payload
                                                  answer:[CMUtilities dictToJsonString:response]
                                     withCompletionBlock:^(BOOL result, NSError *error) {
                                        [LocalStorage addEventToHistory:[NSString stringWithFormat:@"%@ - Answer question", [payloadDict objectForKey:@"question_text"]]];
                                        return completionBlock(result, error);
                                    }];
                                }];
        [alert addAction:button];
    }

    [self presentViewController:alert animated:YES completion:nil];
}

- (void)acceptCredential:(NSString *) data
     withCompletionBlock:(ResponseWithBoolean) completionBlock {
    [CMCredential acceptCredentilaFromMessage:data withCompletionBlock:^(BOOL result, NSError *error) {
        return completionBlock(result, error);
    }];
}

- (void)rejectCredential:(NSString *) data
     withCompletionBlock:(ResponseWithBoolean) completionBlock {
    [CMCredential rejectCredentilaFromMessage:data
                          withCompletionBlock:^(BOOL result, NSError *error) {
        return completionBlock(result, error);
    }];
}

- (void)sendProof:(NSString *) data
withCompletionBlock:(ResponseWithBoolean) completionBlock {
    [CMProofRequest sendProofRequestFromMessage:data
                          withCompletionHandler:^(BOOL result, NSError *error) {
        return completionBlock(result, error);
    }];
}

- (void)rejectProof:(NSString *) data
withCompletionBlock:(ResponseWithBoolean) completionBlock {
    [CMProofRequest rejectProofRequestFromMessage:data
                          withCompletionHandler:^(BOOL result, NSError *error) {
        return completionBlock(result, error);
    }];
}

- (IBAction)addNewConn: (id)sender {
    if(addConnConfigTextView.text.length > 3 && ![addConnConfigTextView.text isEqual: @"enter code here"]) {
        NSDictionary *connectValues = [CMConnection parsedInvite: addConnConfigTextView.text];
        NSString *label = [connectValues objectForKey: @"label"];
        NSString *goal = @"";
        if ([connectValues valueForKey:@"goal"] != nil) {
            goal = [connectValues objectForKey: @"goal"];
        } else {
            goal = @"New connection";
        }
        NSString *profileUrl = [connectValues objectForKey: @"profileUrl"];
        
        NSMutableDictionary* requestsDict = [[LocalStorage getObjectForKey: @"requests" shouldCreate: true] mutableCopy];
        NSString *uuid = [[NSUUID UUID] UUIDString];

        NSDictionary* requestObj = @{
            @"name": label,
            @"profileUrl": profileUrl,
            @"goal": goal,
            @"uuid": uuid,
            @"type": @"null",
            @"data": [CMUtilities dictToJsonString:connectValues]
        };
        [requestsDict setValue: requestObj forKey: uuid];
        [LocalStorage store: @"requests" andObject: requestsDict];
        
        requests = requestsDict;
        [self.tableView reloadData];
    }
}

- (IBAction)checkMessages:(id)sender {
    [CMMessage downloadAllMessages:^(NSArray *responseArray, NSError *error) {
        NSLog(@"downloadAllMessages %@", responseArray);
        for (NSInteger i = 0; i < responseArray.count; i++) {
            NSDictionary *message = responseArray[i];

            NSString *type = [message objectForKey:@"type"];
            NSLog(@"typetypetype %@", type);
            if ([type isEqual:@"credential-offer"]) {
                NSString *payload = [message objectForKey:@"payload"];
                NSArray *payloadArr = [CMUtilities jsonToArray:payload];
                NSDictionary *payloadDict = payloadArr[0];
                
                NSMutableDictionary* requestsDict = [[LocalStorage getObjectForKey: @"requests" shouldCreate: true] mutableCopy];
                NSString *uuid = [[NSUUID UUID] UUIDString];

                NSDictionary* requestObj = @{
                    @"name": [payloadDict objectForKey:@"claim_name"],
                    @"goal": @"Credential offer",
                    @"uuid": [message objectForKey:@"uid"],
                    @"type": type,
                    @"data": [CMUtilities dictToJsonString:message]
                };
                [requestsDict setValue: requestObj forKey: uuid];
                [LocalStorage store: @"requests" andObject: requestsDict];

                dispatch_async(dispatch_get_main_queue(), ^{
                    [CMMessage updateMessageStatus:[message objectForKey:@"pwDid"]
                                         messageId:[message objectForKey:@"uid"]
                               withCompletionBlock:^(BOOL result, NSError *error) {
                        if (error) {
                            NSLog(@"%@", error);
                        }
                    }];
                });

                dispatch_async(dispatch_get_main_queue(), ^{
                    self.requests = requestsDict;
                    [self.tableView reloadData];
                });
            }
            if ([type isEqual:@"committed-question"]) {
                NSString *payload = [message objectForKey:@"payload"];
                NSDictionary *payloadDict = [CMUtilities jsonToDictionary:payload];
                
                NSLog(@"payloadDict oofer %@", payloadDict);
                
                NSMutableDictionary* requestsDict = [[LocalStorage getObjectForKey: @"requests" shouldCreate: true] mutableCopy];
                NSString *uuid = [[NSUUID UUID] UUIDString];

                NSDictionary* requestObj = @{
                    @"name": [payloadDict objectForKey:@"question_text"],
                    @"goal": @"Question",
                    @"uuid": [message objectForKey:@"uid"],
                    @"type": type,
                    @"data": [CMUtilities dictToJsonString:message]
                };
                [requestsDict setValue: requestObj forKey: uuid];
                [LocalStorage store: @"requests" andObject: requestsDict];

                dispatch_async(dispatch_get_main_queue(), ^{
                    [CMMessage updateMessageStatus:[message objectForKey:@"pwDid"]
                                         messageId:[message objectForKey:@"uid"]
                               withCompletionBlock:^(BOOL result, NSError *error) {
                        if (error) {
                            NSLog(@"%@", error);
                        }
                    }];
                });

                dispatch_async(dispatch_get_main_queue(), ^{
                    self.requests = requestsDict;
                    [self.tableView reloadData];
                });
            }
            if ([type isEqual:@"presentation-request"]) {
                NSDictionary *payload = [CMUtilities jsonToDictionary:[message objectForKey:@"payload"]];

                NSLog(@"payloadDict oofer %@", payload);
                
                NSMutableDictionary* requestsDict = [[LocalStorage getObjectForKey: @"requests" shouldCreate: true] mutableCopy];
                NSString *uuid = [[NSUUID UUID] UUIDString];

                NSDictionary* requestObj = @{
                    @"name": [payload objectForKey:@"comment"],
                    @"goal": @"Proof Request",
                    @"uuid": [message objectForKey:@"uid"],
                    @"type": type,
                    @"data": [CMUtilities dictToJsonString:message]
                };
                [requestsDict setValue: requestObj forKey: uuid];
                [LocalStorage store: @"requests" andObject: requestsDict];

                dispatch_async(dispatch_get_main_queue(), ^{
                    [CMMessage updateMessageStatus:[message objectForKey:@"pwDid"]
                                         messageId:[message objectForKey:@"uid"]
                               withCompletionBlock:^(BOOL result, NSError *error) {
                        if (error) {
                            NSLog(@"%@", error);
                        }
                    }];
                });

                dispatch_async(dispatch_get_main_queue(), ^{
                    self.requests = requestsDict;
                    [self.tableView reloadData];
                });
            }
        }
    }];
}

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section {
    return [requests.allKeys count];
}

- (CGFloat)tableView:(UITableView *)tableView heightForRowAtIndexPath:(NSIndexPath *)indexPath
{
   return 110;
}

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath {
    CustomTableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:@"cell" forIndexPath:indexPath];
    if (cell == nil) {
        cell = [[CustomTableViewCell alloc] init];
    }
    
    NSDictionary *requestDict = requests[requests.allKeys[indexPath.row]];
    NSString *type = [requestDict objectForKey: @"type"];
    NSString *name = [requestDict objectForKey: @"name"];
    NSString *goal = [requestDict objectForKey: @"goal"];
    NSString *uuid = [requestDict objectForKey: @"uuid"];
    NSString *data = [requestDict objectForKey: @"data"];

    if ([type isEqual:@"null"]) {
        NSString *logoUrl = [requestDict objectForKey: @"profileUrl"];
        [cell updateAttribute:name
                     subtitle:goal
                      logoUrl:logoUrl
               acceptCallback:^() {
                        [self newConnection:data
                        withCompletionBlock:^(BOOL result, NSError *error) {
                            if (result) {
                                [self switchActionToHistoryView: uuid];
                            }
                        }];
                    }
                rejectCallback:^() {
                    [self switchActionToHistoryView: uuid];
                }
         ];
        [cell.accept setTitle:@"Accept" forState:UIControlStateNormal];
    } else if ([type isEqual:@"credential-offer"]) {
        [cell updateAttribute:name
                     subtitle:goal
                      logoUrl:@""
               acceptCallback:^() {
                         [self acceptCredential:data
                            withCompletionBlock:^(BOOL result, NSError *error) {
                            if (result) {
                                [self switchActionToHistoryView: uuid];
                            };
                        }];
                    }
               rejectCallback:^() {
                        [self rejectCredential:data
                            withCompletionBlock:^(BOOL result, NSError *error) {
                            if (result) {
                                [self switchActionToHistoryView: uuid];
                            };
                        }];
                    }
         ];
        [cell.accept setTitle:@"Accept" forState:UIControlStateNormal];
        [cell.reject setHidden:NO];
    } else if ([type isEqual:@"committed-question"]) {
        [cell updateAttribute:name
                     subtitle:goal
                      logoUrl:@""
               acceptCallback:^() {
                        [self answer:data
                 withCompletionBlock:^(BOOL result, NSError *error) {
                            if (result) {
                                [self switchActionToHistoryView: uuid];
                            };
                        }];
                    }
               rejectCallback:^() {
            [self switchActionToHistoryView: uuid];
               }
         ];
        [cell.accept setTitle:@"Answer" forState:UIControlStateNormal];
    } else if ([type isEqual:@"presentation-request"]) {
        [cell updateAttribute:name
                     subtitle:goal
                      logoUrl:@""
               acceptCallback:^() {
                        [self sendProof:data
                 withCompletionBlock:^(BOOL result, NSError *error) {
                            if (result) {
                                [self switchActionToHistoryView: uuid];
                            };
                        }];
                    }
               rejectCallback:^() {
                    [self rejectProof:data
             withCompletionBlock:^(BOOL result, NSError *error) {
                        if (result) {
                            [self switchActionToHistoryView: uuid];
                        };
                    }];
                }
         ];
        [cell.accept setTitle:@"Accept" forState:UIControlStateNormal];
    }
    return cell;
}

- (void) switchActionToHistoryView:(NSString *) uuid {
    dispatch_async(dispatch_get_main_queue(), ^{
        NSMutableDictionary* processedDict = nil;
        NSDictionary* req = [LocalStorage getObjectForKey: @"requests" shouldCreate: false];
        for (NSInteger i = 0; i < req.allKeys.count; i++) {
            NSString *key = req.allKeys[i];
            if ([key isEqual: uuid]) {
                [processedDict removeObjectForKey:key];
            }
        }
        [LocalStorage store: @"requests" andObject: processedDict];
        self.requests = processedDict;
        [self.tableView reloadData];
    });
}

@end
     

File: examples/ios/CMeSdkSwift/CMeSdkSwift/HomeViewController.swift
//
//  HomeViewController.swift
//  CMeSdkSwift
//
//  Created by Evernym on 06.07.2021.
//  Copyright © 2021 Norman Jarvis. All rights reserved.
//

import AVFoundation
import UIKit
import QRCodeReader

class HomeViewController: UIViewController, QRCodeReaderViewControllerDelegate {
    lazy var readerVC: QRCodeReaderViewController = {
        let builder = QRCodeReaderViewControllerBuilder {
            $0.reader = QRCodeReader(metadataObjectTypes: [.qr], captureDevicePosition: .back)
            
            $0.showTorchButton        = true
            $0.showSwitchCameraButton = true
            $0.showCancelButton       = true
            $0.showOverlayView        = true
            $0.rectOfInterest         = CGRect(x: 0.2, y: 0.2, width: 0.6, height: 0.6)
        }
        
        return QRCodeReaderViewController(builder: builder)
    }()
    
    func reader(_ reader: QRCodeReaderViewController, didScanResult result: QRCodeReaderResult) {
      reader.stopScanning()

      dismiss(animated: true, completion: nil)
    }

    func reader(_ reader: QRCodeReaderViewController, didSwitchCamera newCaptureDevice: AVCaptureDeviceInput) {
        print("Switching capture to: \(newCaptureDevice.device)")
    }

    func readerDidCancel(_ reader: QRCodeReaderViewController) {
      reader.stopScanning()

      dismiss(animated: true, completion: nil)
    }
    
    @IBOutlet var tableView: UITableView!
    @IBOutlet var infoLbl: UILabel!
    @IBOutlet var addConnConfigTextView: UITextView!
    @IBOutlet var newConnLabel: UILabel!
    @IBOutlet var addConnectionButton: UIButton!
    
    var requests: [String: Any] = [:]
    var isInitialized = false;
    
    typealias CompletionHandler = (_ result:Bool, _ error:Error?) -> Void

    override func viewDidLoad() {
        // On iOS diveces you cann't use qr code scnner
        // But you can handle invite with text input and button on home page
        // For show input set this flag to false
        let isHideInput = true;
        
        super.viewDidLoad();
        self.tableView.delegate = self;
        self.tableView.dataSource = self;
        self.addConnConfigTextView.delegate = self;
        self.addConnConfigTextView.layer.cornerRadius = 5;
        
        self.addConnConfigTextView.isHidden = isHideInput;
        self.addConnectionButton.isHidden = isHideInput;
        self.newConnLabel.isHidden = isHideInput;
    }
    
    override func viewWillAppear(_ animated: Bool) {
        super.viewWillAppear(animated);
        if let storageRequests = LocalStorage.getObjectForKey("requests", shouldCreate: false) as? [String: Any] {
            self.requests = storageRequests;
            self.tableView.reloadData();
            NotificationCenter.default.addObserver(self, selector: #selector(self.vcxInitialized), name: NSNotification.Name(rawValue: "vcxInitialized"), object: nil)
            isInitialized = MobileSDK.shared().sdkInited
        }
    }
    
    override func viewWillDisappear(_ animated: Bool) {
        super.viewWillDisappear(animated)
        NotificationCenter.default.removeObserver(self)
    }

    @objc func handleSingleTap(_ sender: UITapGestureRecognizer) {
        self.view.endEditing(true)
    }
    
    @IBAction func scanQR(_ sender: UIButton) {
        if(!isInitialized) {
            print("Please wait for VCX to initialize!")
            return;
        }
        readerVC.delegate = self
        readerVC.completionBlock = { (result: QRCodeReaderResult?) in
            let invite = result?.value as String?
            print(invite ?? "")
            self.addNewConnectionFromQRCode(invite ?? "")
        }
        readerVC.modalPresentationStyle = .formSheet
        present(readerVC, animated: true, completion: nil)
    }
    
    private func addNewConnectionFromQRCode(_ data: String) {
        let connectValues = CMConnection.parsedInvite(data);

        if var storageRequests = LocalStorage.getObjectForKey("requests", shouldCreate: false) as? [String: Any] {
            let label = connectValues?["label"];
            let goal = connectValues?["goal"] ?? "New connection";
            let profileUrl = connectValues?["profileUrl"];
            let uuid = UUID().uuidString
            
            let newRequest:[String:Any] = [
                "name": label as? String ?? "",
                "goal": goal as? String ?? "",
                "profileUrl": profileUrl as? String ?? "",
                "uuid": uuid,
                "type": "null",
                "data": CMUtilities.toJsonString(connectValues)
            ];
            
            storageRequests[uuid] = newRequest;
            LocalStorage.store("requests", andObject: storageRequests);
            
            self.requests = storageRequests;
            self.tableView.reloadData();
        }
        return
    }
    
    
    @objc func vcxInitialized() {
        self.infoLbl.text = "VCX initialized!"
        isInitialized = true;
        let delayInSeconds: Double = 15.0
        DispatchQueue.main.asyncAfter(deadline: .now() + delayInSeconds, execute: {
            self.infoLbl.text = ""
        })
    }
    
    @IBAction func addNewConnection(_ sender: Any) {
        do {
        let connectionText = addConnConfigTextView.text ?? ""
        if connectionText.count > 3 && connectionText != "enter code here" {
            let connectValues = CMConnection.parsedInvite(connectionText);

            if var storageRequests = LocalStorage.getObjectForKey("requests", shouldCreate: false) as? [String: Any] {
                let label = connectValues?["label"];
                let goal = connectValues?["goal"] ?? "New connection";
                let profileUrl = connectValues?["profileUrl"];
                let uuid = UUID().uuidString
//                let data = CMUtilities.dict(toJsonString: connectValues) ?? "";
                
                let messageDictionary = connectValues;
                let jsonData = try JSONSerialization.data(withJSONObject: messageDictionary as Any, options: [])
                let jsonString = String(data: jsonData, encoding: String.Encoding.ascii)!
                
                let newRequest:[String:Any] = [
                    "name": label as? String ?? "",
                    "goal": goal as? String ?? "",
                    "profileUrl": profileUrl as? String ?? "",
                    "uuid": uuid,
                    "type": "null",
                    "data": CMUtilities.toJsonString(connectValues)
                ];
                
                storageRequests[uuid] = newRequest;
                LocalStorage.store("requests", andObject: storageRequests);
                
                self.requests = storageRequests;
                self.tableView.reloadData();
            }
        }
        } catch {
            print(error)
        }
    }
    
    @IBAction func checkMessages(_ sender: Any) {
        CMMessage.downloadAllMessages { messages, error in
            for message in messages ?? [] {
                let msg = message as! [String : String];
                let type = msg["type"];
                
                if type == "credential-offer" {
                    let name = msg["claim_name"]
                    let uid = msg["uid"]
                    
                    if var storageRequests = LocalStorage.getObjectForKey("requests", shouldCreate: false) as? [String: Any] {
                        let uuid = UUID().uuidString;
                        
                        let newRequest = [
                            "name": name ?? "",
                            "goal": "Credential Offer",
                            "uuid": uuid,
                            "type": type ?? "",
                            "data": CMUtilities.dict(toJsonString: message as? [AnyHashable : Any]) ?? ""
                        ] as [String:String?]?;
                        
                        storageRequests[uuid] = newRequest;
                        LocalStorage.store("requests", andObject: storageRequests);
                        
                        if (msg["pwDid"] != nil) && (msg["uid"] != nil) {
                            CMMessage.updateStatus(msg["pwDid"]!, messageId: msg["uid"]!) { _,_ in  }
                        }
                        
                        self.requests = storageRequests;
                        self.tableView.reloadData();
                    }
                }
                if type == "committed-question" {
                    let payload = msg["payload"];
                    let payloadDict = CMUtilities.json(toDictionary: payload);

                    if var storageRequests = LocalStorage.getObjectForKey("requests", shouldCreate: false) as? [String: Any] {
                        let uuid = UUID().uuidString
                        
                        let newRequest:[String:String] = [
                            "name": payloadDict?["question_text"] as? String ?? "",
                            "goal": "Question",
                            "uuid": uuid,
                            "type": type ?? "",
                            "data": CMUtilities.dict(toJsonString: message as? [AnyHashable : Any]) ?? ""
                        ];
                        
                        storageRequests[uuid] = newRequest;
                        LocalStorage.store("requests", andObject: storageRequests);
                        
                        if (msg["pwDid"] != nil) && (msg["uid"] != nil) {
                            CMMessage.updateStatus(msg["pwDid"]!, messageId: msg["uid"]!) { _,_ in  }
                        }
                        
                        self.requests = storageRequests;
                        self.tableView.reloadData();
                    }
                }
                if type == "presentation-request" {
                    let payload = msg["payload"];
                    let payloadDict = CMUtilities.json(toDictionary: payload);

                    if var storageRequests = LocalStorage.getObjectForKey("requests", shouldCreate: false) as? [String: Any] {
                        let uuid = UUID().uuidString
                        
                        let newRequest:[String:String] = [
                            "name": payloadDict?["comment"] as? String ?? "",
                            "goal": "Proof Request",
                            "uuid": uuid,
                            "type": type ?? "",
                            "data": CMUtilities.dict(toJsonString: message as? [AnyHashable : Any]) ?? ""
                        ];
                        
                        storageRequests[uuid] = newRequest;
                        LocalStorage.store("requests", andObject: storageRequests);
                        
                        if (msg["pwDid"] != nil) && (msg["uid"] != nil) {
                            CMMessage.updateStatus(msg["pwDid"]!, messageId: msg["uid"]!) { _,_ in  }
                        }
                        
                        self.requests = storageRequests;
                        self.tableView.reloadData();
                    }
                }
            }
        }
    }
    
    @objc private func newConnection(_ data: String, completionHandler: @escaping CompletionHandler) -> Any? {
//        CMConnection.handle(_:data, connectionType:ConnectionType.QR.rawValue, phoneNumber:"") { result, error in
//            if error != nil {
//                return completionHandler(false, error);
//            }
//            return completionHandler(true, nil);
//        }
        
        CMConnection.handle(data, connectionType:ConnectionType.QR.rawValue, phoneNumber:"") { result, error in
            if error != nil {
                return completionHandler(false, error);
            }
            return completionHandler(true, nil);
        };
        return completionHandler(false, nil);
    }
    
    private func acceptCredential(_ data: String, completionHandler: @escaping CompletionHandler) -> Any? {
        CMCredential.acceptCredentila(fromMessage:data) { result, error in
            if error != nil {
                return completionHandler(false, error);
            }
            return completionHandler(true, nil);
        }
    }
    
    private func rejectCredential(_ data: String, completionHandler: @escaping CompletionHandler) -> Any? {
        CMCredential.rejectCredentila(fromMessage:data) { result, error in
            if error != nil {
                return completionHandler(false, error);
            }
            return completionHandler(true, nil);
        }
    }
    
    private func sendProof(_ data: String, completionHandler: @escaping CompletionHandler) -> Any? {
        CMProofRequest.send(fromMessage:data) { result, error in
            if error != nil {
                return completionHandler(false, error);
            }
            return completionHandler(true, nil);
        }
    }
    
    private func rejectProof(_ data: String, completionHandler: @escaping CompletionHandler) -> Any? {
        CMProofRequest.reject(fromMessage:data) { result, error in
            if error != nil {
                return completionHandler(false, error);
            }
            return completionHandler(true, nil);
        }
    }
    
    func convertToDictionary(text: String) -> [String: Any]? {
        if let data = text.data(using: .utf8) {
            do {
                return try JSONSerialization.jsonObject(with: data, options: []) as? [String: Any]
            } catch {
                print(error.localizedDescription)
            }
        }
        return nil
    }
    
    private func answer(_ data: String, completionHandler: @escaping CompletionHandler) -> Any? {
        let msg = convertToDictionary(text: data)
        let payload = msg?["payload"];
        let payloadDict = convertToDictionary(text: payload as! String)
        let responses = payloadDict?["valid_responses"] as? [[String : String]];
        
        let pwDidMes = msg?["pwDid"];
        let questionConnection = CMConnection.getByPwDid(pwDidMes as? String);
        
        let alert = UIAlertController(
            title: payloadDict?["question_text"] as? String,
            message: payloadDict?["question_detail"] as? String,
            preferredStyle: UIAlertController.Style.alert
        );
        
        for response in responses ?? [] {
            alert.addAction(
                UIAlertAction(
                    title: response["text"],
                    style: UIAlertAction.Style.default
                ) {_ in
                    CMMessage.answerQuestion(questionConnection!, message: payload as! String, answer: CMUtilities.dict(toJsonString: response)) { result, error in
                        LocalStorage.addEvent(toHistory: NSString.localizedStringWithFormat("%@ - Answer question", payloadDict?["question_text"] as! CVarArg) as String);
                        return completionHandler(result, error);
                    }
                }
            )
        }
        self.present(alert, animated: true)
        return completionHandler(false, nil);
    }
    
    private func requestByIndex(_ index: Int) -> Any? {
        let requestsIDs = Array(requests.keys);
        if requestsIDs.count > 0 {
            let requestsID = requestsIDs[index];
            return requests[requestsID];
        }
        return [:]
    }
    
    private func switchRequestToHistoryView(_ uuid: String, completionHandler: @escaping CompletionHandler) -> Any? {
        var storageRequests = requests;
        print("switchRequestToHistoryView", storageRequests, uuid)
        for key in Array(storageRequests.keys) {
            if key == uuid {
                storageRequests.removeValue(forKey: key);
            }
        }
        LocalStorage.store("requests", andObject: storageRequests);
        self.requests = storageRequests;
        self.tableView.reloadData();
        return completionHandler(true, nil);
    }
    
    private func addRejectAction(_ name: String) {
        LocalStorage.addEvent(toHistory: "\(name) - Rejected")
        self.tableView.reloadData();
        return
    }
}

extension HomeViewController: UITextViewDelegate, UITableViewDelegate, UITableViewDataSource {
    func tableView(_ tableView: UITableView, heightForRowAt indexPath: IndexPath) -> CGFloat {
        return 110;
    }
    
    func numberOfSections(in tableView: UITableView) -> Int {
        return 1;
    }
    
    func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        return requests.count;
    }
    
    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        let cell: CustomTableViewCell = tableView.dequeueReusableCell(withIdentifier: "cell", for: indexPath) as! CustomTableViewCell;
        let request = self.requestByIndex(indexPath.row) as? [String: Any] ?? [:];
        
        let type = request["type"] as? String ?? "";
        let name = request["name"] as? String ?? "";
        let goal = request["goal"] as? String ?? "";
        let uuid = request["uuid"] as? String ?? "";
        let data = request["data"] as? String ?? "";

        if (type == "null") {
            let logoPath = request["logoUrl"] as? String ?? "";
            cell.updateCellAttributes(
                title: name,
                subtitle: goal,
                logoUrl: logoPath);
            cell.addAceptCallback(acceptCallback: { () -> () in
                _ = self.newConnection(data) { result, _ in
                    if (result) {
                        _ = self.switchRequestToHistoryView(uuid) { _, _ in }
                    }
                }
            });
            cell.addRejectCallback(rejectCallback: { () -> () in
                _ = self.switchRequestToHistoryView(uuid) { _, _ in }
                _ = self.addRejectAction(uuid)
            });
        }
        if (type == "credential-offer") {
            cell.updateCellAttributes(
                title: name,
                subtitle: goal,
                logoUrl: "");
            cell.addAceptCallback(acceptCallback: { () -> () in
//                _ = self.switchRequestToHistoryView(uuid) { _, _ in };
                _ = self.acceptCredential(data) { result, error in
                    _ = self.switchRequestToHistoryView(uuid) { _, _ in };
                }
            });
            cell.addRejectCallback(rejectCallback: { () -> () in
                _ = self.rejectCredential(data) { result, error in
                    if (result) {
                        _ = self.switchRequestToHistoryView(uuid) { _, _ in }
                    }
                }
            });
        }
        if (type == "committed-question") {
            cell.updateCellAttributes(
                title: name,
                subtitle: goal,
                logoUrl: "");
            cell.addAceptCallback(acceptCallback: { () -> () in
                _ = self.answer(data) { result, error in
                    if (result) {
                        _ = self.switchRequestToHistoryView(uuid) { _, _ in };
                    }
                }
            });
            cell.addRejectCallback(rejectCallback: { () -> () in
                _ = self.switchRequestToHistoryView(uuid) { _, _ in }
            });
        }
        if (type == "presentation-request") {
            cell.updateCellAttributes(
                title: name,
                subtitle: goal,
                logoUrl: "");
            cell.addAceptCallback(acceptCallback: { () -> () in
                _ = self.sendProof(data) { result, error in
                    if (result) {
                        _ = self.switchRequestToHistoryView(uuid) { _, _ in };
                    }
                }
            });
            cell.addRejectCallback(rejectCallback: { () -> () in
                _ = self.rejectProof(data) { result, error in
                    if (result) {
                        _ = self.switchRequestToHistoryView(uuid) { _, _ in }
                    }
                }
            });
        }
        return cell;
    }
}
